ipld/amt/src/diff.rs

```diff
@@ -0,0 +1,387 @@
+// Copyright 2021-2023 Protocol Labs
+// Copyright 2019-2022 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use anyhow::Context;
+use fvm_ipld_blockstore::Blockstore;
+use fvm_ipld_encoding::CborStore;
+use serde::{de::DeserializeOwned, Serialize};
+
+use crate::node::CollapsedNode;
+
+use super::*;
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum ChangeType {
+    Add,
+    Remove,
+    Modify,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub struct Change<V> {
+    pub change_type: ChangeType,
```

Comment:
The change type doesn't seem necessary given that we have a before/after, right? If we want to make it convenient, we can have helper methods like:

```rust
fn is_new(&self) -> bool {
    self.before.is_none() && self.after.is_some()
}
```

Reply:
How about removing the field from the struct and keep the enum for the helper function?
```
impl<V> Change<V> {
    pub fn change_type(&self) -> ChangeType {
        match (&self.before, &self.after) {
            (Some(_), Some(_)) => ChangeType::Modify,
            (Some(_), None) => ChangeType::Remove,
            (None, Some(_)) => ChangeType::Add,
            (None, None) => panic!("Invalid change type, before and after cannot be both None"),
        }
    }
}
```